hack/tools/release/notes.go

```diff
@@ -403,8 +446,82 @@ REPLACE ME: A couple sentences describing the deprecation, including links to do
 
 	fmt.Println("")
 	fmt.Println("_Thanks to all our contributors!_ 😊")
+}
 
-	return 0
+func parseWeeklyUpdate(commits []*commit, merges map[string][]string, commitRange string, lastDay string) {
+	for _, c := range commits {
+		body := trimTitle(c.body)
+		var key, prNumber, fork string
+		switch {
+		case strings.HasPrefix(body, ":sparkles:"), strings.HasPrefix(body, "✨"):
+			key = features
+			body = strings.TrimPrefix(body, ":sparkles:")
+			body = strings.TrimPrefix(body, "✨")
+		case strings.HasPrefix(body, ":bug:"), strings.HasPrefix(body, "🐛"):
+			key = bugs
+			body = strings.TrimPrefix(body, ":bug:")
+			body = strings.TrimPrefix(body, "🐛")
+		case strings.HasPrefix(body, ":book:"), strings.HasPrefix(body, "📖"):
+			key = documentation
+			body = strings.TrimPrefix(body, ":book:")
+			body = strings.TrimPrefix(body, "📖")
+			if strings.Contains(body, "CAEP") || strings.Contains(body, "proposal") {
+				key = proposals
+			}
+		case strings.HasPrefix(body, ":seedling:"), strings.HasPrefix(body, "🌱"):
+			key = other
+			body = strings.TrimPrefix(body, ":seedling:")
+			body = strings.TrimPrefix(body, "🌱")
+		case strings.HasPrefix(body, ":warning:"), strings.HasPrefix(body, "⚠️"):
+			key = warning
+			body = strings.TrimPrefix(body, ":warning:")
+			body = strings.TrimPrefix(body, "⚠️")
```

Comment:
Thanks for the PR @willie-yao! I gave it a try running locally and something like:

```
go run hack/tools/release/notes.go weekly --since 2023-08-13 --until 2023-09-06 --milestone v1.6
```

outputs:

```
Weekly update :rotating_light:
Changes from 2023-08-13 to 2023-09-06 a total of 108 new commits where merged into main.

- 4 Breaking Changes :warning:
         - Remove v1alpha3 API Version (#8997)
         - : Remove go-vcs dependency from releaselink tool (#9288)
         - Bump to controller-runtime v0.16 (#8999)
         - Improve Context handling in clusterctl (#8939)

- 6 New Features :sparkles:
         - Add MachinePool workers support in ClusterClass (#9016)
         - Add automation to create release branch and tags (#9111)
         - block move with annotation (#8690)
         - machineset: adjust preflight check to allow kubelet version skew of 3 for clusters running v1.28 and above (#9222)
         - v1.28: Prepare quickstart, capd and tests for the new release including kind bump (#9160)
         - Add topology-owned label to MachineHealthChecks. (#9191)

- 12 Bug Fixes :bug:
         - Fix autoscaler image repo (#9353)
         - fix go install path for kpromo v4.0.4 (#9336)
         - MD controller: use regular random suffix for MachineSets, ensure max length 63 (#9298)
         - Deduplicating area in pr title in release notes (#9186)
         - Adding metrics container port in tilt-prepare only if it's missing (#9308)
         - Fix tilt-prepare leader-elect setting (#9315)
         - Allow duplicate objects in Tiltfile (#9302)
         - test: pin conformance image to a version which includes a fix for the dualstack tests (#9252)
         - tilt: ensure .tiltbuild/bin directory is created early enough, add tilt troubleshooting guide (#9165)
         - Certificate paths in cloud-init scripts should not use a platform-dependent path separator (#9167)
         - Allow to drop useExperimentalRetryJoin field from KubeadmControlPlane.kubeadmConfigSpec (#9170)
         - Fix AddAnnotations for unstructured.Unstructured (#9164)

- 76 Other changes :seedling:

- 10 Documentation and book contributions :book: 

All merged PRs can be viewed in GitHub:
https://github.com/kubernetes-sigs/cluster-api/pulls?q=is%3Apr+closed%3A2023-08-13..2023-09-07+is%3Amerged+milestone%3Av1.6+

_Thanks to all our contributors!_ 😊
/Your friendly comms release team
```

I wonder if we should also consider a case of PR titles like in https://github.com/kubernetes-sigs/cluster-api/pull/9288 PR in the code or just fix it manually 😄 

Anyways, to have a proper output for just :warning: PRs (it could happen to any PR category), this change:
```suggestion
		case strings.HasPrefix(body, ":warning:"), strings.HasPrefix(body, "⚠️"):
			key = warning
			parts := strings.Split(body, ":")
			if len(parts) > 1 {
				body = strings.TrimSpace(parts[len(parts)-1])
			} else {
				body = strings.TrimPrefix(body, ":warning:")
				body = strings.TrimPrefix(body, "⚠️")
				body = strings.TrimSpace(body)
			}
```

should suffice, but again, we might have more than `:` or any other unwanted/unpredictable char in the PR titles due to typos/by overlooking 🙂  So my intent here is, we either consider covering the corner uses cases like the above and extend the code a bit or just ignore it since it is not a big deal and always can be fixed manually.

Reply:
Thanks for the suggestion @furkatgofurov7! I think we should be diligent to catch these errors and I feel like there may be too many edge cases to account for all possible typos. However I do agree that it would be nice to have it be fixed automatically. Wdyt @g-gaston @mjlshen @nprokopic?

Reply:
Yeah, I could see something like ":warning: - PR title" being used as well. Not sure if any other characters make sense other than ":" and "-", but I'm ok with letting them fall through for now as I think the notes still make sense.